### ISSUE TITLE
No Flowdock templates in 1.1-alpha4.54.105

### ISSUE THREAD

**Issue by rlaneve** [tcplugins/tcWebHooks#20]
We just updated to version 1.1-alpha4.54.105 in order to use the built-in support for Flowdock. After installing the plugin and restarting the server, we don't have Flowdock as an option in the payload formats list. Checking the TC server log, we're seeing the following errors when the plugin starts up:

```
[2015-12-07 12:58:43,321]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildStarted-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,321]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,322]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildSuccessful-buildFixed-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,322]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,322]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildSuccessful-buildFixed-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,322]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,323]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-responsibilityChanged-normal.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,323]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildStarted-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,323]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,324]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildSuccessful-buildFixed-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,324]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,324]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildSuccessful-buildFixed-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,324]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-buildBroken-buildFailed-buildInterrupted-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,325]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: An Error occurred trying to load the template file: webhook/teamcity/payload/template/flowdock\flowdock-responsibilityChanged-branch.json. The file was not found in the classpath. 
[2015-12-07 12:58:43,325]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: Failed to register template flowdock. No regular template configurations were found. 
[2015-12-07 12:58:43,378]  ERROR -   jetbrains.buildServer.SERVER - FlowdockWebHookTemplate :: Failed to register template flowdock. No branch template configurations were found. 
```

None of the other templates had an issue - they all loaded successfully.

If there's any additional info we can provide to help troubleshoot this, let me know, and thanks for the great plugin!

**Comment by netwolfuk** [comment]
Thanks for your report. 
What version of tcwebhooks did you upgrade from? A 1.1 or a 0.9? 
Also what OS is the server on? It looks like windows. 
What version of TeamCity?

**Comment by rlaneve** [comment]
We upgraded from tcwebhooks 0.9.29.63.
We're running TeamCity 9.1.1, on Windows 2008 R2.

**Comment by netwolfuk** [comment]
I must confess I've not tested on windows much. I'll give it a test. If I don't get a new release out by tomorrow, it might be later in the week as I'm traveling.

**Comment by rlaneve** [comment]
No problem; thanks for looking into it.
Just an FYI, we do have a couple of spaces in path of our TeamCity data directory. The logs regarding the plugin only referenced the relative path, so just wanted to mention it in case it helps; in case something's not being escaped somewhere.

**Comment by netwolfuk** [comment]
Ah good point. I'll test with a space in the path too. If you hang around Java long enough you get out of that habit ;-)

**Comment by netwolfuk** [comment]
Can you try this build. It works for me. I was able to reproduce on windows (without a space in the path).

https://teamcity.jetbrains.com/repository/download/WebHooksAndOtherPlugins_TcWebHooksBranch09xX/644405:id/tcWebHooksPlugin-1.1.55.106.zip

Just login as guest if you need to. If it works for you, I'll push it as a new alpha release for 1.1

**Comment by rlaneve** [comment]
That fixed it!

**Comment by netwolfuk** [comment]
Great! I'll organise a release sometime this week.
I'd be keen to hear any feedback you have on the templates and how they work for you in Flowdock. 
How could they be better, or do they do what you are expecting in your flows?
Thanks!

**Comment by rlaneve** [comment]
It's looking good so far. There are a couple of things I'm not sure about, but let me preface by saying that we are using this latest version as a replacement for some TeamCity notifications in Flowdock that we put together ourselves using the 0.9.x version of the tcWebHooks plugin. We've been seeing those notifications for several years, so I might just need to get used to the new ones.

As a point of reference, here's an example of what our prior notifications looked like:
![tc-flowdock-custom](https://cloud.githubusercontent.com/assets/37863/11658987/ed5d3b52-9d93-11e5-9941-945fc0096a7d.png)
We only sent notifications when a build finished (success or failure), and we could see who kicked off the build (the avatar of the notification), the name of the TC project, the name of the TC build config, and the TC build number ("BKR CRM", "Dev", and "2733-34" respectively in the example).

The equivalent using the latest plugin looks like this:
![tc-flowdock-new1](https://cloud.githubusercontent.com/assets/37863/11659052/39dc08dc-9d94-11e5-9741-17e93f74b973.png)
It's not as easy for us to see the TC project name (it's in a tag, but we'd prefer it in the notification's "subject" line). We also can't see the build number or user without clicking the notification to view the details. We can probably live with that.

The threading is great, and we've enabled both the "Build Started" and "Build Finished" notifications thanks to the new plugin's threading support. However, it seems to be re-using the same thread for too many builds. Take a look at this example, and the details:
![tc-flowdock-new2](https://cloud.githubusercontent.com/assets/37863/11659117/86bea2fe-9d94-11e5-9869-92507a6913f4.png)
![tc-flowdock-new2-detail](https://cloud.githubusercontent.com/assets/37863/11659126/932d1f8e-9d94-11e5-94bb-aad985825dfa.png)

In that example, we ran two builds for the same project, same build config, but 17 hours apart, and each of the two builds had a unique build number. We would have preferred those two to show up as separate threads. I'm not sure if you have that much control over the threading, though, or whether any other Flowdock users would agree with us on this point.

Again, the new plugin is great, and thanks again for putting in the effort and sharing!

**Comment by netwolfuk** [comment]
Thanks for your detailed reply.

I think we could put something together that merges the best of both our approaches. I'm not a flowdock user. So I was just throwing together a first cut at the templates to be improved upon.

I'm impressed with what you've achieved in the 0.9 plugin. There's some serious XML hacking there. :-) Could you post the old webhook configuration somewhere?

Changing anything in the template is fine. If you have a look at it you'll see how easy it is. The thread is just a string in the template.

I'm happy to be challenged on the threading. My reasoning was that each build type will get it's own thread. So i based the thread id on a combination of build name (and branch name if applicable). Otherwise you'd end up with thousands of threads. I guess I was thinking they are a sort of channel and you can see the build history for a build in that channel. The latest build status is always reflected in the state of the thread. But maybe flows are a better metaphor for a channel and threads are valid per instance if a build. I just don't like seeing lots of red threads when later that build is actually fixed.

Text in the subject is easy to change too. Happy to add whatever you want in there. I suspect my build names in TeamCity are more explicit than yours. Eg. Mine is not just "Dev" ;-)

**Comment by netwolfuk** [comment]
Here the link to the one of the templates. The thread id is highlighted.
https://github.com/tcplugins/tcWebHooks/blob/custom-templates/tcwebhooks-core/src/main/resources/webhook/teamcity/payload/template/flowdock/flowdock-buildSuccessful-buildFixed-branch.json#L8

You can see all the templates here:
https://github.com/tcplugins/tcWebHooks/blob/custom-templates/tcwebhooks-core/src/main/resources/webhook/teamcity/payload/template/flowdock/

And to change the "subject" it's this line.
  "thread": {
  "title": "${buildName} (${sanitize(branchDisplayName)})",

That should include ${projectName}

Would you be interested in putting together some changes to the templates and then sending through a pull request?

**Comment by netwolfuk** [comment]
I'm really interested in how you obtained the user/avatar. I'll await your old webhooks config.

**Comment by netwolfuk** [comment]
BTW, the end game is to have a nice way for you to customise the templates. Then you could build whatever you wanted on top of the default templates.
Here is my current thinking https://github.com/tcplugins/tcWebHooks/issues/8#issuecomment-124078127
It's technically possible to overwrite a template with customised content if you're up for XML hacking, although I need to fix issue #18

**Comment by rlaneve** [comment]
For our original integration, we didn't customize anything in tcWebHooks or TeamCity. We spun up a small, internal API, and we configured the web hooks in TC to send to that API. The API then re-formatted the data and sent it over to Flowdock, using one of their original (now deprecated) integration APIs. It wasn't this one exactly, but the formatting is very similar to the even-older-one we're still using: https://www.flowdock.com/api/team-inbox
That end-point allows specifying the sender's name and email address, and when doing so Flowdock automatically displays the related avatar. Again though, we're not real concerned with losing info about who triggered the build - it's available in the thread detail, and that's good enough for us.

At work, my team is responsible for building internal systems. Right now, we've got about 30 projects setup in TeamCity - one per "system" (think timesheet, expense reimbursement, commissions, CRM, etc...) - and within each TC project there are typically two builds, always named "Dev" (i.e., "develop") and "Release". So, in our environment, we definitely need to see both the TC project name and the TC build config name to know what's actually being built. Additionally, in the last year or so we've started building GitHub pull requests via the "Dev" config in some projects, so in those cases we will need to see the TC "branch name" as well.

Considering the above, we come to the threading issue. If a single thread is used for each build within a project+buildconfig+branch combination, for most of our projects that means we'd only ever have a couple of threads (one for the "Dev" build+branch, and one for the "Release" build+branch). Some of these projects have been around a long time, and/or have frequent builds. In our CRM project, for the main "Dev" build config, right now we'd have over 5,400 notifications in the thread for that project+buildconfig+branch combination. I imagine we'd be waiting awhile if we ever tried to view the detail on that thread.  ;)

I'm totally fine with customizing the templates on our end; this is the definitely the kind of thing that everyone is gonna wanna tweak in one way or another. I'll take a look at the two issues you referenced and see what I can do on our TeamCity server as soon as I get a chance.

**Comment by netwolfuk** [comment]
Aha. Yes that all makes sense. Thanks for taking the time to explain.
For elastic support I added a ${date(simpledateformat)} variable. We could actually group threads by week or day or month if that would be useful.

But what really needs to happen is for me to finish the template editor ui and fix the template loading bug. Then ask the power is in your hands.

**Comment by netwolfuk** [comment]
Yey. Fixed #18 
It was a weird Spring instance issue.

**Comment by rlaneve** [comment]
Sounds great. Looking forward to trying it out and tweaking some templates.

**Comment by netwolfuk** [comment]
The crux of this issue is that templates should be user modifiable and/or creatable.
This is now being addressed by #67 so I will close this issue.
